src/main/scala/is/hail/io/bgen/BgenWriter.scala

```diff
@@ -0,0 +1,186 @@
+    val totalFractional = resized.map(d => d._2 - d._3).sum
+
+    resized.sortBy(d => -(d._2 - d._3)).zipWithIndex.map { case (r, i) =>
+      if (i < totalFractional)
+        (r._1, math.ceil(r._2))
+      else
+        (r._1, math.floor(r._2))
+    }.sortBy(_._1).map(_._2.toUInt)
+  }
+
+  def emitHeaderBlock(vds: VariantDataset): Array[Byte] = {
+    val nSamples = vds.nSamples
+    val nVariants = vds.countVariants()
+    assert(nVariants <= math.pow(2, 31) - 1, s"Maximum number of variants can export is (2^31 - 1). Found $nVariants.")
+
+    val magicNumbers = Array("b", "g", "e", "n").flatMap(_.getBytes)
+    val flags = 0x01 | (0x02 << 2) | (0x01 << 31)
+    val sampleIdentifierBlock = emitSampleIdentifierBlock(vds.sampleIds.toArray)
+    val headerLength = 20
+    val sampleIdentifierLength = sampleIdentifierBlock.length
+    val offset = headerLength + sampleIdentifierLength
+
+    val byteBuffer = new ArrayBuilder[Byte]
+
+    byteBuffer ++= intToBytesLE(offset)
+    byteBuffer ++= intToBytesLE(headerLength)
+    byteBuffer ++= intToBytesLE(nVariants.toInt)
+    byteBuffer ++= intToBytesLE(nSamples)
+    byteBuffer ++= magicNumbers
+    byteBuffer ++= intToBytesLE(flags)
+
+    byteBuffer.result() ++ sampleIdentifierBlock
+  }
+
+  def emitSampleIdentifierBlock(sampleIds: Array[String]): Array[Byte] = {
+    val byteBuffer = new ArrayBuilder[Byte]
+    sampleIds.foreach(s => byteBuffer ++= stringToBytesWShortLength(s))
+
+    val sampleIdBytes = byteBuffer.result()
+    val sampleBlockLength = sampleIdBytes.length + 8
+    val nSamples = sampleIds.length
+
+    intToBytesLE(sampleBlockLength) ++ intToBytesLE(nSamples) ++ sampleIdBytes
+  }
+
+  def emitGenotypeData(gs: Iterable[Genotype], samplePloidies: Array[Byte],
+    nAlleles: Int, nBitsPerProb: Int, conversionFactor: Double, bitMask: Long): (Array[Byte], Int, Int) = {
+    val byteBuffer = new ArrayBuilder[Byte]
+    val compresser = new Deflater()
+
+    val sampleProbs = new ArrayBuilder[UInt]
+    val nGenotypes = triangle(nAlleles)
+    val noCall = Array.fill[UInt](nGenotypes - 1)(UInt(0))
+
+    gs.zipWithIndex.foreach { case (gt, i) =>
+      val px = gt.unboxedPX
+      if (px == null) {
+        samplePloidies(i) = 0x80.toByte
+        sampleProbs ++= noCall
+      }
+      else {
+        samplePloidies(i) = ploidy
+        sampleProbs ++= resizeProbInts(px, conversionFactor).dropRight(1)
+      }
+    }
+
+    byteBuffer ++= intToBytesLE(samplePloidies.length)
+    byteBuffer ++= shortToBytesLE(nAlleles)
+    byteBuffer += ploidy
+    byteBuffer += ploidy
+    byteBuffer ++= samplePloidies
+    byteBuffer += phased
+    byteBuffer += nBitsPerProb.toByte
+    byteBuffer ++= bitPack(sampleProbs.result(), nBitsPerProb)
+
+    val uncompressedData = byteBuffer.result()
+    compresser.setInput(uncompressedData)
+    compresser.finish()
+
+    val outputStream = new ByteArrayOutputStream(uncompressedData.length)
```

Comment:
I wouldn't involve ByteArrayOutputStream here.  Also, uncompressed length with be unnecessarily big.  You should just grow an array and copy it to make the size right.